packages/glob/src/internal-pattern-helper.ts

```diff
@@ -0,0 +1,142 @@
+import * as pathHelper from './internal-path-helper'
+import {IGlobOptions} from './internal-glob-options'
+import {MatchResult} from './internal-match-result'
+import {Pattern} from './internal-pattern'
+
+const IS_WINDOWS = process.platform === 'win32'
+
+/**
+ * Given an array of patterns, returns an array of paths to search.
+ * Duplicates and paths under other included paths are filtered out.
+ */
+export function getSearchPaths(patterns: Pattern[]): string[] {
+  // Ignore comment and negate patterns
+  patterns = patterns.filter(x => !x.comment && !x.negate)
+
+  // Create a map of all search paths
+  const searchPathMap: {[key: string]: string} = {}
+  for (const pattern of patterns) {
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    searchPathMap[key] = 'candidate'
+  }
+
+  const result: string[] = []
+
+  for (const pattern of patterns) {
+    // Check if already included
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    if (searchPathMap[key] === 'included') {
+      continue
+    }
+
+    // Check for an ancestor search path
+    let foundAncestor = false
+    let tempKey = key
+    let parent = pathHelper.dirname(tempKey)
+    while (parent !== tempKey) {
+      if (searchPathMap[parent]) {
+        foundAncestor = true
+        break
+      }
+
+      tempKey = parent
+      parent = pathHelper.dirname(tempKey)
+    }
+
+    // Include the search pattern in the result
+    if (!foundAncestor) {
+      result.push(pattern.searchPath)
+      searchPathMap[key] = 'included'
+    }
+  }
+
+  return result
+}
+
+/**
+ * Matches the patterns against the path
+ */
+export function match(patterns: Pattern[], itemPath: string): MatchResult {
+  let result: MatchResult = MatchResult.None
+
+  for (const pattern of patterns) {
+    if (pattern.comment) {
+      continue
+    }
+
+    if (pattern.negate) {
+      result &= ~pattern.match(itemPath)
+    } else {
+      result |= pattern.match(itemPath)
+    }
+  }
+
+  return result
+}
+
+/**
+ * Parses the pattern strings into Pattern objects
+ */
+export function parse(patterns: string[], options: IGlobOptions): Pattern[] {
+  const result: Pattern[] = []
+
+  for (let patternString of patterns) {
+    // Parse
+    const pattern = new Pattern(patternString)
```

Comment:
Seems like it would be better to check if the `pattern` is a comment before creating the `Pattern` instance, that way the class itself doesn't have to worry about comments and you can remove some of the extra checks in `internal-pattern.ts`

Reply:
i like this way for encapsulation (parsing is the responsibility of the Pattern class)

Reply:
the internal-pattern-helper.parse function never actually returns the comment patterns to the caller, so most of the \"is comment\" filters are not actually required. i just put them there so it's clear.

Reply:
unresolve if you disagree

Reply:
A comment isn't a `Pattern` though, so it seems weird to make it one. Maybe a factory or function to create Patterns that can also return `undefined` for comments would be a good middle-ground, and then you could get rid of the nullable/unknown fields in `Pattern` that are required for only comments.

Reply:
> A comment isn't a Pattern

sold. refactored and i like it much better. thanks